### ISSUE TITLE
Redirection error causes theme coding error

### ISSUE THREAD

**Issue by agrowe** [catalyst/moodle-auth_saml2#425]
**What happened?**

When accessing the course category index page (`/course/index.php?categoryid=77`), Totara will show the below error message:

```
PHP message: Default exception handler: Coding error detected, it must be fixed by a programmer: page layout file [dirroot]/theme/basis/layout/default.php does not contain the main content placeholder, please include "<?php echo $OUTPUT->main_content() ?>" in theme layout file.
```
```
Debug: \n
Error code: codingerror\n
line 840 of /lib/outputrenderers.php: coding_exception thrown\n
line 77 of /course/index.php: call to core_renderer->header()
```

Version of auth_saml2: 2020031800
Version of Totara: 12.12 (Moodle 3.3)
SAML configs:

- all users set to ldap auth
- Dual login is passive
- Allowed any auth type is yes
- SAML is configured to ADFS on prem

**Testing Steps**

1. Open a new window without being signed into IDP
2. Paste in the url directly to `/course/index.php?categoryid=77` for a category that exists
3. SAML will redirect to our ADFS for auth
4. Redirected back to the `/course/index.php?categoryid=77` page but above error message is displayed
5. If you refresh the page or reload the same URL, it will take you to the category page without the error message. It is only on the first load after being authenticated via SAML

**What does work**

Following same steps above, each time in a new incognito window

- Paste in URL to site home page, auth will prompt and redirect to home page without issue
- Paste in URL to a course page `/course/view.php?id=123`, auth will prompt and redirect to the course page without issue
- Paste in the course index URL but without a category id, `/course/index.php`, auth will prompt and redirect to the Totara catalog without issue.
- Disable SAML, go to the `/course/index.php?categoryid=77` URL, redirect to login page, auth via LDAP and redirect to category page without issue

**Comment by agrowe** [comment]
Updates with SAML debugging turned on. The SAML response has been removed as it authenticates properly each time and would add too much to the logs.

**DOESN'T WORK: Going to the standard category index page and the coding exception is shown**

New incognito window, going straight to Totara URL `/course/index.php?categoryid=77`

```
PHP message: auth_saml2: pre_loginpage_hook enter
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: pre_loginpage_hook enter
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: saml_login_complete to lowercase for 0 => user
PHP message: auth_saml2: saml_login_complete found user user
PHP message: auth_saml2: saml_login_complete continuing onto https://[site]/course/index.php?categoryid=77
PHP message: auth_saml2: pre_loginpage_hook exit
PHP message: Default exception handler: Coding error detected, it must be fixed by a programmer: page layout file [dirroot]/theme/basis/layout/default.php does not contain the main content placeholder, please include "&lt;?php echo $OUTPUT-&gt;main_content() ?&gt;" in theme layout file. Debug: 
Error code: codingerror
 line 840 of /lib/outputrenderers.php: coding_exception thrown
 line 77 of /course/index.php: call to core_renderer->header()
```

**WORKS: Going to the standard course view page and everything works**

New incognito window, going straight to Totara URL `/course/view.php?id=236`

```
PHP message: auth_saml2: pre_loginpage_hook enter
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: pre_loginpage_hook enter
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: saml_login_complete to lowercase for 0 => user
PHP message: auth_saml2: saml_login_complete found user user
PHP message: auth_saml2: saml_login_complete continuing onto https://[site]/course/view.php?id=236
PHP message: auth_saml2: pre_loginpage_hook exit
```

**WORKS: Going to the category index page but this time adding saml=off**

New incognito window, going straight to Totara URL `/course/index.php?categoryid=77&saml=off`

```
PHP message: auth_saml2: pre_loginpage_hook enter
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect skipping due to saml=off parameter
PHP message: auth_saml2: loginpage_hook exit
PHP message: auth_saml2: pre_loginpage_hook exit
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: loginpage_hook enter
PHP message: auth_saml2: should_login_redirect enter
PHP message: auth_saml2: should_login_redirect redirecting due to passive mode.
PHP message: auth_saml2: saml_login_complete to lowercase for 0 => user
PHP message: auth_saml2: saml_login_complete found user user
PHP message: auth_saml2: saml_login_complete redirecting to https://[site]/course/index.php?categoryid=77
```

**Comment by agrowe** [comment]
Further debugging to show the issue is related to multiple sessions being created which don't match to the required sesskey when generating the theme layout.

Around Line 825 in `/lib/outputrenderers.php` for Totara V12 (might be different in Moodle) the following block of code is run for the `header()` function.

```
// Find the appropriate page layout file, based on $this->page->pagelayout.
$layoutfile = $this->page->theme->layout_file($this->page->pagelayout);
// Render the layout using the layout file.
$rendered = $this->render_page_layout($layoutfile);

// Slice the rendered output into header and footer.
$cutpos = strpos($rendered, $this->unique_main_content_token);
if ($cutpos === false) {
	$cutpos = strpos($rendered, self::MAIN_CONTENT_TOKEN);
	$token = self::MAIN_CONTENT_TOKEN;
} else {
	$token = $this->unique_main_content_token;
}

if ($cutpos === false) {
	throw new coding_exception('page layout file ' . $layoutfile . ' does not contain the main content placeholder, please include "<?php echo $OUTPUT->main_content() ?>" in theme layout file.');
}
```

I added the below code after the call to `$rendered = $this->render_page_layout($layoutfile);` to log what token it was trying to find as well as the token the `header()` function was expecting.

```
file_put_contents("/tmp/rendered.html", $rendered);
error_log($this->unique_main_content_token);
```

After the exception occurs:
The unique main content token was: `[MAIN CONTENT GOES HERE - Ch4p1EkecX]`
BUT, the token in the logged rendered.html file was: `[MAIN CONTENT GOES HERE - 1i1iFLPyOE]`

When refreshing the page and going back to the course index page with a categoryid, the session would then be active and authenticated so the sesskey would be correct.

**Comment by agrowe** [comment]
Totara have included a fix for this in their next releases of 12.19, 11.28, 10.34, and 9.45 with a tracker of TL-25216.

The issue was due to the theme renderer being initialised before the sesskey was created as part of require_login and the SAML plugin authenticating users without redirecting back to the same page. 

This is why it worked with saml=off as it forced a redirect rather than allowing the script to continue.

Totara have removed the sesskey requirement from the renderer and used a static unique token instead, but the SAML plugin could also be edited so that it redirects after authentication back to the originally requested location.
